lib/src/settings.rs

```diff
@@ -146,3 +174,37 @@ impl UserSettings {
         &self.config
     }
 }
+
+/// This Rng uses interior mutability to allow generating random values using an
+/// immutable reference. It also fixes a specific seedable RNG for
+/// reproducibility.
+#[derive(Debug)]
+pub struct JJRng(Mutex<ChaCha20Rng>);
+impl JJRng {
+    /// Wraps Rng::gen but only requires an immutable reference.
+    pub fn gen<T>(&self) -> T
+    where
+        rand::distributions::Standard: rand::distributions::Distribution<T>,
+    {
+        let mut rng = self.0.lock().unwrap();
+        rng.gen()
+    }
+
+    /// Creates a new RNGs. Could be made public, but we'd like to encourage all
+    /// RNGs references to point to the same RNG.
+    fn new(seed: Option<u64>) -> Self {
+        Self(Mutex::new(JJRng::internal_rng_from_seed(seed)))
+    }
+
+    fn reset(&self, seed: Option<u64>) {
+        let mut rng = self.0.lock().unwrap();
+        *rng = JJRng::internal_rng_from_seed(seed)
```

Comment:
Requiring `&mut self` is safer (because consumer of RNG shouldn't be allowed to reset the seed.) This also means we don't have to deal with shared RNG, and we can just recreate `JJRng`.

```
self.rng = Arc::new(JJRng::new(new_rng_seed))`
```

Reply:
I tried that, but it's not doable as long as we use an `Arc`:

```
error[E0596]: cannot borrow data in an `Arc` as mutable
  --> lib/src/settings.rs:79:13
   |
79 |             self.rng.reset(new_rng_seed);
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot borrow as mutable
   |
   = help: trait `DerefMut` is required to modify through a dereference, but it is not implemented for `std::sync::Arc<settings::JJRng>`
```

Reply:
I didn't want to create a new `Arc`, again, because I want every use of an RNG to reference the same RNG.

Reply:
Ah, got it, thanks. I didn't consider such level of guarantee. I just thought calling `&mut UserSettings` function or recreating `UserSettings` means we don't care any existing state.

(Feel free to mark this as resolved.)